### ISSUE TITLE
let_and_return lint false positive

### ISSUE THREAD

**Issue by alexschrod** [rust-lang/rust-clippy#5729]
I've encountered a false positive with the `let_and_return` lint. Here's a minimum example that demonstrates the issue: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=37bcec97617120eafab94ca5216bf579

If you run clippy on it, you get the following output:
```
warning: returning the result of a `let` binding from a block
  --> src/lib.rs:18:9
   |
17 |         let clone = Arc::clone(&self.foo);
   |         ---------------------------------- unnecessary `let` binding
18 |         clone
   |         ^^^^^
   |
   = note: `#[warn(clippy::let_and_return)]` on by default
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#let_and_return
help: return the expression directly
   |
17 |         
18 |         Arc::clone(&self.foo)
   |
```
If you try to follow its advice, however, and try to return `Arc::clone(&self.foo)` directly, you will instead get this compilation error:
```
error[E0308]: mismatched types
  --> src/lib.rs:17:20
   |
13 | impl<T: Foo + 'static> FooStorage
   |      - this type parameter
...
17 |         Arc::clone(&self.foo)
   |                    ^^^^^^^^^ expected trait object `dyn Foo`, found type parameter `T`
   |
   = note: expected reference `&std::sync::Arc<dyn Foo>`
              found reference `&std::sync::Arc<T>`
   = help: type parameters must be constrained to match other types
   = note: for more information, visit https://doc.rust-lang.org/book/ch10-02-traits.html#traits-as-parameters
```
